### PR TITLE
Add `Runtime::enter`

### PR DESCRIPTION
**Pull request by bryanburgers**
Add a `Runtime::enter` function that allows the caller to enter the runtime context without calling `run`, `spawn`, or `block_on`. This mirrors the `enter` function from Tokio 0.2's runtime.

---

There were three approaches I saw to #24.

1. Expose the underlying tokio_02::Runtime
2. Expose a tokio_02::Handle
3. Expose an `enter` function. I opted for this one.

All three have the question of what to do with the idle tracker (I did nothing 😬). By implementing an `enter`, this also guarantees that everything is set up on the tokio_01 compatibility layer, I think.

This solves the problem in #24 that I need solved, using the multi-threaded runtime. I tested it on our app and it works perfectly.

However, I don't know enough about the current_thread runtime to get the same test to pass.

### CODE REVIEW

File: src/runtime/current_thread/runtime.rs
Comment (hawkw) [inline src/runtime/current_thread/runtime.rs]: I think we need to add a call to `with_idle` here, in order for the runtime's idle tracking to work.
```suggestion
        executor_01::with_default(&mut spawner, &mut enter, |_enter| Self::with_idle(idle, move || inner.enter(f)))
```

File: src/runtime/current_thread/runtime.rs
Comment (hawkw) [inline src/runtime/current_thread/runtime.rs]: I notice this code is _very_ similar to `Runtime::run`:
https://github.com/tokio-rs/tokio-compat/blob/867b1a5bacedcd2aa715d5cb68b59eb889ae136b/src/runtime/current_thread/runtime.rs#L357-L375 . 
and `Runtime::block_on_std`:
https://github.com/tokio-rs/tokio-compat/blob/867b1a5bacedcd2aa715d5cb68b59eb889ae136b/src/runtime/current_thread/runtime.rs#L335-L351
Could we consider refactoring those methods to be implemented using `enter`? The code duplication seems like a potential source of errors, since we would need to reflect changes in a number of places.

File: tests/rt_current_thread.rs
Comment (hawkw) [inline tests/rt_current_thread.rs]: I think the reason this test is failing is because we are `spawn`ing the future. This means it is now running _on the `tokio-compat` runtime_. The call to `tokio` 0.1's `spawn` function returns a `Result`. `Result` implements `IntoFuture` by returning a future that finishes _immediately_ with the `Ok` or `Err` value, so when we call `into_future`, we get back a completed `Ok(())` future. Then, we `block_on` that future, which also returns immediately. However, the _spawned future_, which is running on the `tokio-compat` runtime, is never run, since we never actually run that runtime's futures (such as with a call to `Runtime::block_on` or `Runtime::run`/`shutdown_on_idle`).

I think if the test was changed to something like this, it would pass:
```rust
    let future_ran = Arc::new(AtomicBool::new(false));
    let ran = future_ran.clone();
    let rt = current_thread::Runtime::new().unwrap();
    rt.enter(|| {
        let future = futures_01::future::lazy(move || {
            future_ran.store(true, Ordering::SeqCst);
            Ok(())
        });
        futures_03::executor::block_on(future)
    }).unwrap();
    assert!(ran.load(Ordering::SeqCst));
```
However, this wouldn't really test the behavior of `enter`, since the same code would work with the `tokio_compat` runtime totally removed. I think what we would really want to be testing is whether code inside of the `enter` can spawn futures on the entered runtime that _will_ be run when that runtime is driven. We might want to have a test like this:
```rust
    let future1_ran = Arc::new(AtomicBool::new(false));
    let ran1 = future1_ran.clone();
    let future2_ran = Arc::new(AtomicBool::new(false));
    let ran2 = future2_ran.clone();
    let rt = current_thread::Runtime::new().unwrap();
    rt.enter(|| {
        let future = futures_01::future::lazy(move || {
            tokio_01::spawn(futures_01::future::lazy(move || {
                future2_ran.store(true, Ordering::SeqCst);
                Ok(())
            }))?;    
            future1_ran.store(true, Ordering::SeqCst);
            Ok(())
        });
        futures_03::executor::block_on(future)
    }).unwrap();
    assert!(ran1.load(Ordering::SeqCst));

    // The runtime should run any futures spawned in the `enter` block.
    rt.run();
    assert!(ran2.load(Ordering::SeqCst));
```

I'm not convinced the `futures_03::executor::block_on` is strictly necessary for this test, so I think we can simplify this to just something like this:
```rust
    let future_ran = Arc::new(AtomicBool::new(false));
    let ran = future_ran.clone();
    let rt = current_thread::Runtime::new().unwrap();
    rt.enter(|| {
        tokio_01::spawn(futures_01::future::lazy(move || {
            future_ran.store(true, Ordering::SeqCst);
            Ok(())
        })).unwrap();
    });

    // The runtime should run any futures spawned in the `enter` block.
    rt.run();
    assert!(ran1.load(Ordering::SeqCst));
```
Does that make sense?

File: tests/rt_threadpool.rs
Comment (hawkw) [inline tests/rt_threadpool.rs:195]: Same comments as my suggestions on the `current_thread` version of this test. Additionally, I think we might want to consider having a test where a threadpool runtime's `Handle` is entered on _another_ thread, while the runtime is driven on the main thread, and we spawn a future from inside the `enter` and assert that it runs?

File: src/runtime/current_thread/runtime.rs
Comment (hawkw) [inline src/runtime/current_thread/runtime.rs:414]: So, this looks like the `enter` method on `tokio` 0.2's runtime, and that will work in some cases. However, `enter` requires us to be able to immutably borrow the runtime. This means that we cannot, for example, use `enter` on an external thread while we are running the runtime using `run` or `block_on` on a different thread.

I think we will also want to add an `enter` method to the `Handle` struct, like [`tokio` 0.2 has](https://docs.rs/tokio/0.2.9/tokio/runtime/struct.Handle.html#method.enter). That way, we can write code like this:
```rust
let rt = current_thread::Runtime::new().unwrap();
let handle = rt.handle().clone();
let thread1 = std::thread::spawn(move || {
    handle.enter(|| {
        futures::executor::block_on(async { // or whatever
            tokio::spawn(async {
                // ... whatever ...
            }).await;
            // ...
        });
    });
});

rt.block_on(async {
    // ...
});
thread1.join().unwrap();
```
This would probably necessitate some changes to the compat runtime's `Handle` type.

I think we could then implement `Runtime::enter` as
```rust
pub fn enter<F, R>(&self, f: F) -> R
where
    F: FnOnce() -> R,
{
    self.handle().enter(f)
}
```

Of course, this applies to the threadpool runtime as well.

**Review by hawkw — CHANGES_REQUESTED** [review]
Thanks for working on this. I had some thoughts. Let me know if you have any questions!

File: tests/rt_threadpool.rs
Comment (bryanburgers) [reply tests/rt_threadpool.rs:195]: OK, I have both tests reliably working right now.

File: tests/rt_current_thread.rs
Comment (bryanburgers) [reply tests/rt_current_thread.rs]: Yeah, that makes sense. For whatever reason, I thought [`Spawn`](https://docs.rs/tokio/0.1.19/tokio/executor/struct.Spawn.html) was something that could be awaited. It can't.

File: src/runtime/current_thread/runtime.rs
Comment (bryanburgers) [reply src/runtime/current_thread/runtime.rs]: Yep, I'll see if that makes sense. At the very least, we can extract out the common code into a non-public method.

File: src/runtime/current_thread/runtime.rs
Comment (bryanburgers) [reply src/runtime/current_thread/runtime.rs]: Done.

File: src/runtime/current_thread/runtime.rs
Comment (bryanburgers) [reply src/runtime/current_thread/runtime.rs:414]: Absolutely agree. So we'd have `enter` on the threadpool's `TaskExecutor`, which is serving the purpose of a `Handle`.

I'll take a look at this in the next couple of days.

**Review by bryanburgers — COMMENTED** [review]
Thanks for your feedback!

I made a little bit of progress on this. Hopefully I can keep chipping away in the next couple of days.

File: tests/rt_current_thread.rs
Comment (hawkw) [reply tests/rt_current_thread.rs]: In `tokio` 0.2, [it is](https://docs.rs/tokio/0.2.9/tokio/task/index.html#spawning), but not in `tokio` 0.1 :)

File: src/runtime/current_thread/runtime.rs
Comment (hawkw) [reply src/runtime/current_thread/runtime.rs:414]: For the threadpool, I would probably add a `Handle` type _in addition_ to the `TaskExecutor`, since `Handle` is an 0.2 API while `TaskExecutor` is an 0.1 API. I'd like to put the 0.2 methods (like `enter`) on a type with the same name as the 0.2 type (`Handle`), so that it's easier for users to transition from `tokio-compat` to using `tokio` 0.2 directly. Just my 2 cents.

Comment (hawkw) [reply src/runtime/current_thread/runtime.rs:414]: FWIW, I think we can add this stuff to the `Handle` types in a follow-up PR if you'd like to get this change merged sooner...

File: tests/rt_threadpool.rs
Comment (hawkw) [reply tests/rt_threadpool.rs:195]: 🎉

File: tests/rt_threadpool.rs
Comment (hawkw) [inline tests/rt_threadpool.rs]: I don't think this import is still necessary?

File: tests/rt_current_thread.rs
Comment (hawkw) [inline tests/rt_current_thread.rs]: I don't think this import is still necessary.

File: Cargo.toml
Comment (hawkw) [inline Cargo.toml]: I don't think this dependency is needed, now that we've removed the calls to `futures_03::executor::block_on` from the tests.

**Review by hawkw — COMMENTED** [review]
I commented on a few things that ought to be cleaned up. Once those are addressed, I'm happy to merge this, if you'd like to make the `Handle` changes in a separate follow-up PR; alternatively, I am also happy to wait until those changes have been made in this branch — whatever you prefer!

**Review by hawkw — APPROVED** [review]
great, LGTM! let's see about adding this to the `Handle` types as well in a follow-up?

**Pull request merged**
